html/inc/boinc_db.inc

```diff
@@ -27,119 +27,93 @@ incs();
-                    $instance = null;
+                    //error_log("BoincDb::get_aux(): connected to replica DB $dbnum");
                     self::$instance = $instance;
-                    return $instance;
-                } else {
-                    error_log("BoincDb::get_aux(): Couldn't connect to $user@$replica_host (trying $user@$host next)");
+                    return;
                 }
             }
+            // if can't connect to replica, fall through and try DB 0
         }
-        $retval = $instance->init_conn($user, $passwd, $host, $name, false);
-        if (!$retval) {
-            $instance = null;
-            error_log("BoincDb::get_aux(): Couldn't connect to $user@$host (giving up)");
-        } else {
-            // needed for places where we do direct queries
-            if (!$instance->do_query("use $name")) {
-                error_log("BoincDb::get_aux(): Couldn't select database $name on $host (giving up)");
-                $instance = null;
-            }
+        $host = parse_config($config, '<db_host>');
+        if (!$host) $host = 'localhost';
+        $user = parse_config($config, '<db_user>');
+        $name = parse_config($config, '<db_name>');
+        $passwd = parse_config($config, '<db_passwd>');
+        if (!$name || !$user || !$passwd) {
+            error_log("BoincDb::get_aux(): must specify DB name, user, passwd");
+            return;
+        }
+        $retval = $instance->init_conn($user, $passwd, $host, $name);
+        if ($retval) {
+            //error_log("BoincDb::get_aux(): connected to DB $dbnum");
+            self::$instance = $instance;
+            return;
         }
-        self::$instance = $instance;
-        return $instance;
+        error_log("BoincDb::get_aux(): Couldn't connect to DB $dbnum");
     }
 
-    // same, but
+    // connect to DB $dbnum, but first:
     // 1) check for a cached connection
     // 2) check whether the "stop_web" trigger file is present
     //
-    static function get($readonly = false, $fallback_mode = 0) {
+    // If there's a page that's guaranteed to do only reads, put
+    // BoincDb::get(true);
+    // at the top of it.
+    //
+    // Note: true == 1.
+    // You can also use 2, 3... to select other replicas
+    //
+    static function get($dbnum = 0) {
         global $generating_xml;
-        if (!isset(self::$instance)) {
-            if (web_stopped()) {
-                if ($generating_xml) {
-                    xml_error(-183, "project down for maintenance");
-                } else {
-                    show_project_down();
-                }
-            }
-            self::get_aux($readonly, $fallback_mode);
-            if (!self::$instance) {
-                if ($generating_xml) {
-                    xml_error(-138, "Can't connect to database");
-                } else {
-                    error_page("Can't connect to database");
-                }
+        if (isset(self::$instance)) {
```

Comment:
Hm, this implies that there can only be a single instance at a time. What we need here is a hash of instances (indexed by `$dbnum`) such that multiple instances are retained and/or can be switched between. E.g. the server status page needs to query different databases (here: replicas).

Reply:
I think a hash of instances might be a bit of an overkill, I think E@H is the only project that uses as much as two. However a requirement for us would be to access different replicas at least in sequence. I did this in d1f99043877552ea48ac2912dc4ec9ac9a1efddd

Reply:
Ok, but please add a comment that states that there's only ever be one current DB instance at a time. IOW, the current instance gets created, reused or replaced by a new one.

Reply:
Is this a requirement?

Reply:
Well no, but that's how you implemented it, isn't it?

Reply:
Resolved [here](https://github.com/BOINC/boinc/pull/5695#discussion_r1698695506)